Migrate option parsing from legacy methods to newopts.shlib

TakJim opened the pull request:
## Summary
Migrates ACPCAlignment.sh and AnatomicalAverage.sh from legacy option parsing methods to the standardized newopts.shlib system used in the HCP pipeline.

## Changes Made
### ACPCAlignment.sh
- ✅ Converted `getopt1()` function calls to `opts_AddMandatory/opts_AddOptional`
- ✅ Added all RIKEN-specific options with proper defaults
- ✅ Preserved brain extraction and BET parameter functionality
- ✅ Improved error handling and validation

### AnatomicalAverage.sh
- ✅ Unified option parsing using newopts.shlib
- ✅ Maintained backward compatibility with positional arguments
- ✅ Enhanced validation (minimum 2 images required)
- ✅ Improved logging and configuration display

## Backward Compatibility
✅ All existing functionality preserved
✅ Legacy argument formats still supported where applicable
✅ Default values maintained

## Related Issues
Resolves merge conflicts between HEAD and RIKEN branches for option parsing standardization.

glasserm commented:
These diffs are not making sense to me.  They seem incomplete relative to the changes mentioned.

coalsont commented:
> These diffs are not making sense to me. They seem incomplete relative to the changes mentioned.

I have changed the base so it is trying to pull into `master` instead of a throwaway branch.  Things should make more sense now.

coalsont commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -1,9 +1,11 @@
+#  installed versions of: FSL5.0.1+
+#  environment: FSLDIR
 
+<<<<<<< HEAD
```
Conflict to clean up.

TakJim commented:
@coalsont 
Thank you for pointing that out. I apologize not all of the merge conflicts have been resolved yet. This pull request is still a work in progress and currently in Draft status. Once I’ve addressed the remaining issues, I’ll reopen it for review.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -171,6 +171,12 @@ then
     #fix this if the script is more than one level below HCPPIPEDIR
     export HCPPIPEDIR="$(dirname -- "$0")/.."
 fi
+# parameter initialization
+InitDof="${InitDof:=""}"
+TXwTemplateBrain="${TXwTemplateBrain:=""}"
+OutputTXwImageSTRING="${OutputTXwImageSTRING:=""}"
+CustomMask="${CustomMask:=""}"
+betbiasfieldcor="${betbiasfieldcor:=""}"
```
Is this temporary?  We don't use environment variables to pass pipeline-specific settings in any other pipelines.

coalsont commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -26,7 +26,59 @@ source "$HCPPIPEDIR/global/scripts/newopts.shlib" "$@"
 
 opts_SetScriptDescription "Script for using topup to do distortion correction for EPI (scout)"
 
+<<<<<<< HEAD
```
Remaining conflict.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -463,7 +526,26 @@ if [ "$CustomBrain" = "NONE" ] ; then
             --in=${wdir}/${TXwImage}${i} \
             --out=${TXwFolder}/${TXwImage}${i}_gdc \
             --owarp=${TXwFolder}/xfms/${TXwImage}${i}_gdc_warp
+<<<<<<< HEAD
```
Remaining conflict.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -396,6 +426,38 @@ fi
+
+  # -- Then we are using existing images
+  else
+    log_Msg "Skipping all the steps preceding AtlasRegistration, using existing images instead. This overrides the option of --runmode."
```
This else doesn't appear to override `RunMode`, though its message says it does.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,419 @@
+	TopupConfig="${HCPPIPEDIR_Config}/b02b0.cnf" #Config for topup or "NONE" if not used
+	UnwarpDir="${UnwarpDir:-z}" # z appears to be best or "NONE" if not used, read from Read.direction in Seriesinfo.txt (PA:y, AP:y-, RL:x, LR:x-, FH:z, HF:z-)
+	BiasFieldSmoothingSigma="5.0"
+	T1wTemplate="${HCPPIPEDIR_Templates}/MNI152_T1_0.7mm.nii.gz"  
```
Sometimes these are 0.8mm or even 1mm.

coalsont replied:
I would consider passing arguments to the `source` line - variables crossing through a script boundary are a bit surprising (we make exceptions for things that nearly all scripts need, like HCPPIPEDIR).  It would look something like this:

```
source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh "$Species" "$StrucRes"
```

TakJim replied:
I’ve reflected your ideas in https://github.com/Washington-University/HCPpipelines/pull/350/commits/ddd4e0c9371ca41082f541c738393bea947f061a .
I modified PreFreeSurferPipelineBatchNHP.sh to use option-based arguments, in the same way as PreFreeSurferPipelineBatch.sh.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,419 @@
+	bettop2center="30"           # distance in mm from the top of FOV to the center of brain in robustroi
+
+	BrainTemplate="Mac30BS"
+	T1wTemplate="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/MNINonLinear/${BrainTemplate}_AverageT1w_restore.nii.gz" 
```
Should we name these templates consistently with the resolution given that for macaques there are both 0.5mm and 0.32mm templates?

TakJim replied:
@takuya-hayashi 
Do you have any ideas on a revision policy regarding the resolution of the T1wTemplate and the naming of the Mac30BS template, in order to take multiple resolutions into account?

takuya-hayashi replied:
How about defining 'StrucResolution' or like before exporting this SetUpSPECIES.sh? (e.g., StrucResolution=0.7 for HCP-YA; StrucResolution=0.5 for macaque default; StrucResolution=0.32 for high-res macaque)

TakJim replied:
I modified `PreFreeSurferPipelineBatchNHP.sh` so that StrucRes is set as an environment variable. https://github.com/Washington-University/HCPpipelines/commit/ddd4e0c9371ca41082f541c738393bea947f061a .

https://github.com/Washington-University/HCPpipelines/pull/350#discussion_r2434396012

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -502,46 +603,134 @@ if [ "$CustomBrain" = "NONE" ] ; then
       fi
 
-      # ACPC align T1w or T2w image to specified MNI Template to create native volume space
-      log_Msg "Aligning ${TXw} image to ${TXwTemplate} to create native volume space"
```
Why are we losing these log messages?

TakJim replied:
Sorry, I accidentally deleted the log message. I fixed it in 41f778850a0d008cfeebcc92508e

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -502,46 +603,134 @@ if [ "$CustomBrain" = "NONE" ] ; then
-        --brainsize=${BrainSize}
-
-      # Brain Extraction(FNIRT-based Masking)
-      log_Msg "Performing Brain Extraction using FNIRT-based Masking"
```
I prefer this log message to the one that replaced it.

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -502,46 +603,134 @@ if [ "$CustomBrain" = "NONE" ] ; then
+	--workingdir=${TXwFolder}/BrainExtraction_FNIRTbased \
+	--in=${TXwFolder}/${TXwImage}_acpc \
+	--ref=${TXwTemplate} \
+	--refmask=${AtlasSpaceFolder}/TemplateMask \
```
I don't see anything in the code above supporting this change.

TakJim replied:
I've reverted it in b379b205c66073e9d829bfbc9bdc62e0992d09ca .

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -550,6 +739,12 @@ if [ "$CustomBrain" = "NONE" ] ; then
 
+   if [ $(imtest ${T2wFolder}/T2w) = 1 ] ; then 	 # added T2w as a phase zero volume - TH Jan 2023
+        SpinEchoPhaseEncodeZero=${T2wFolder}/T2w
+   fi
```
This is a new behavior.  Perhaps it should be optional and configurable? For example, if the T2w scan is actually a FLAIR, this might not work well.

TakJim replied:
Consulted with @takuya-hayashi and added the --flair option. When set to TRUE, the process is skipped. Also modified the logic so that it applies conditionally to NHP only.
https://github.com/Washington-University/HCPpipelines/pull/350/commits/2443c7646e1203b7daf1c0a2699090ee9407020a

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -642,6 +842,8 @@ if [ "$CustomBrain" = "NONE" ] ; then
     log_Msg "Performing Bias Field Correction using T1w image only"
+    BiasFieldFastSmoothingSigma="20"
+    BiasFieldFastSmoothingSigma="--bfsigma=${BiasFieldFastSmoothingSigma}"
 
```
Perhaps this flag should be used in the invocation, rather than being stored in a variable.

TakJim replied:
@takuya-hayashi
what do you think about this idea mentioned in the above comment?

coalsont replied:
Oh, I see, when T1w and T2w are available, you use...**4.0** mm in Chimp (per SetUpSPECIES), but when there is no T2w, you use 20mm.  This still seems like a per-session quirk that the batch should handle somehow, rather than something to hardcode into the pipeline itself.

TakJim replied:
Thank you for the detailed feedback.
For the NHP pipeline, I added BiasFieldFastSmoothingSigma as an optional parameter via opts_AddOptional, with its default value set to 20.
For Human data, it shares the same value as BiasFieldSmoothingSigma.
164ca337b31a05d3de0e60ed53aea9f589f1e9a8

coalsont replied:
I was wondering whether the batch should instead check whether T2w images exist, and if not, change the value of the ordinary bias field sigma, rather than introducing a second one.

If we do want it to be a separate argument, it should probably be named something like `--bias-field-sigma-no-T2w`, if I understand its purpose correctly.

glasserm replied:
Rename BiasFieldFastSmoothingSigma --> BiasFieldSmoothingSigmaNoT2w

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -749,9 +952,9 @@ ${RUN} ${HCPPIPEDIR_PreFS}/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh \
-  --refmask=${TemplateMask} \
+  --ref=${AtlasSpaceFolder}/T1wTemplate \
+  --refbrain=${AtlasSpaceFolder}/T1wTemplateBrain \
+  --refmask=${AtlasSpaceFolder}/TemplateMask \
```
As above, not sure what supports this change in the code.

TakJim replied:
@takuya-hayashi
could you explain the intention behind the directory change in light of the above comment?

TakJim replied:
Following our Friday discussion, we decided to revert the change, and I have made the corresponding update in fc0545d573d163addf884fc2e2caa310cf607e03.

glasserm commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -84,37 +106,84 @@ echo " " >> $WD/log.txt
-verbose_echo " --> Croping the FOV"
-${FSLDIR}/bin/robustfov -i "$Input" -m "$WD"/roi2full.mat -r "$WD"/robustroi.nii.gz $BrainSizeOpt
+fslmaths "$Reference2mm" "$WD"/Reference
+fslmaths "$Reference2mm" -mas "$Reference2mmMask" "$WD"/ReferenceBrain
```
Are these always needed, or just for some of the modes?  We shouldn't make these if we don't need them.

TakJim replied:
@takuya-hayashi 
How about adding an NHP-specific branch so that these files are not created when they are unnecessary?

TakJim replied:
I've add ! Human Conditional in e0c7520ef6e8f6aba2eb107b26ba503fca0948c3.

glasserm commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -84,37 +106,84 @@ echo " " >> $WD/log.txt
+
+verbose_echo " --> Geting a 6 DOF approximation"
 # Concatenate matrices to get full FOV to MNI
-verbose_echo " --> Concatenating matrices to get full FOV to MNI"
```
Another log message that is disappearing.

glasserm commented on PreFreeSurfer/scripts/AnatomicalAverage.sh:
```diff
@@ -66,7 +70,8 @@ if [ X$wdir = X ] ; then
     if [ -f $wdir ] ; then 
-		log_Err_Abort "A file already exists with the name $wdir - cannot use this as the working directory"
+	echo "A file already exists with the name $wdir - cannot use this as the working directory"
+	exit 1;
```
I don't understand the need for this change.

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -1,8 +1,9 @@
-#  installed versions of: FSL
-#  environment: HCPPIPEDIR, FSLDIR
+#  installed versions of: FSL5.0.1+
+#  environment: FSLDIR
```
As above, why are we changing these?

TakJim replied:
@takuya-hayashi
would it be okay to revert this change, or is it a necessary one?

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -80,16 +83,14 @@ opts_ShowValues
-verbose_red_echo " ===> Running Atlas Registration to MNI152"
-verbose_echo " "
+echo " "
+echo " START: AtlasRegistration to MNI152"
```
Again, why are we removing/changing logging messages?

TakJim replied:
I fixed it in 10b02191571bca3999c1209457504aa6702e1ee9

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -99,43 +100,37 @@ echo "PWD = `pwd`" >> $WD/xfms/log.txt
+ ${FSLDIR}/bin/fnirt --in=${T1wRestore} --ref=${Reference2mm} --aff=${WD}/xfms/acpc2MNILinear.mat --refmask=${Reference2mmMask} --fout=${OutputTransform} --jout=${WD}/xfms/NonlinearRegJacobians.nii.gz --refout=${WD}/xfms/IntensityModulatedT1.nii.gz --iout=${WD}/xfms/2mmReg.nii.gz --logout=${WD}/xfms/NonlinearReg.txt --intout=${WD}/xfms/NonlinearIntensities.nii.gz --cout=${WD}/xfms/NonlinearReg.ni ...
 
 # Input and reference spaces are the same, using 2mm reference to save time
-verbose_echo " --> Computing 2mm warp"
```
Again with removing logging messages.

TakJim replied:
Sorry, I accidentally deleted the log message. I fixed it in 10b02191571bca3999c1209457504aa6702e1ee9

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -99,43 +100,37 @@ echo "PWD = `pwd`" >> $WD/xfms/log.txt
 
 # T2w set of warped outputs (brain/whole-head + restored/orig)
 if [ ! "${T2wImage}" = "NONE" ] ; then
-  verbose_echo " --> Creating T2w set of warped outputs"
```
Again with removing logging messages.

TakJim replied:
Sorry, I accidentally deleted the log message. I fixed it in 10b02191571bca3999c1209457504aa6702e1ee9

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -99,43 +100,37 @@ echo "PWD = `pwd`" >> $WD/xfms/log.txt
-  ${FSLDIR}/bin/applywarp --rel --interp=nn -i ${T2wRestoreBrain} -r ${Reference} -w ${OutputTransform} -o ${OutputT2wImageRestoreBrain}
-  ${FSLDIR}/bin/fslmaths ${OutputT2wImageRestore} -mas ${OutputT2wImageRestoreBrain} ${OutputT2wImageRestoreBrain}
-else
-  verbose_echo " ... skipping T2w processing"
```
Again with removing logging messages.

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -99,43 +100,37 @@ echo "PWD = `pwd`" >> $WD/xfms/log.txt
-
-log_Msg "END: AtlasRegistration to MNI152"
+echo " "
+echo " END: AtlasRegistration to MNI152"
```
Again with removing logging messages.

glasserm commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -444,12 +501,30 @@ else
+    --oT2im=${WD}/${T2wImageBasename}_restore \
+    --oT2brain=${WD}/${T2wImageBasename}_restore_brain \
+    ${BiasFieldSmoothingSigma}
+
```
Has any registration occurred between the T1w and T2w at this point?

TakJim replied:
@glasserm This part is already applied only for NHP. Is this okay?


https://github.com/Washington-University/HCPpipelines/blob/c3ccf76fcfd66074cb81403df0f451070d52cfd6/PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh#L503-L521

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -155,15 +259,76 @@ if [ ! $GradientDistortionCoeffs = "NONE" ] ; then
-  # Merge both sets of images
-  ${FSLDIR}/bin/fslmerge -t ${WD}/BothPhases ${WD}/PhaseOne_gdc ${WD}/PhaseTwo_gdc
-
+  ${FSLDIR}/bin/fslmaths ${WD}/PhaseOne_mask_gdc -mas ${WD}/PhaseTwo_mask_gdc -bin ${WD}/Mask  # -ero is too aggressive for NHP - TH 2024
```
Should this be a conditional then?

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -257,7 +497,26 @@ if [[ $UnwarpDir = [xyij] ]] ; then
+   ${GlobalScripts}/bet4animal ${WD}/SBRef ${WD}/SBRef_brain -m -z $species -c $centerx $centery $centerz -f 0.4
+    ${GlobalScripts}/bet4animal ${WD}/PhaseTwo_gdc_one ${WD}/PhaseTwo_gdc_one_brain -m -z $species -c $centerx $centery $centerz -f 0.4
+  fi
+  ${FSLDIR}/bin/flirt -dof 6 -interp spline -in ${WD}/SBRef_brain -ref ${WD}/PhaseTwo_gdc_one_brain -omat ${WD}/SBRef2PhaseTwo_gdc.mat -out ${WD}/SBRef2PhaseTwo_gdc -nosearch
```
Since this is a behavior change, should it be conditional on NHP?

glasserm replied:
@takuya-hayashi please respond to this.

TakJim replied:
I've add ! Human conditional in 93a6fd9daf82d21c8df4912df025c3aaef501d7d.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -363,13 +386,16 @@ log_Msg "T2wFolder: $T2wFolder"
 log_Msg "AtlasSpaceFolder: $AtlasSpaceFolder"
 
 # Unpack List of Images
-T1wInputImages=`echo ${T1wInputImages} | sed 's/@/ /g'`
-T2wInputImages=`echo ${T2wInputImages} | sed 's/@/ /g'`
+T1wInputImages=`echo ${T1wInputImages} | sed 's/@/ /g' | sed -e  's/^[ \t]*//'`
+T2wInputImages=`echo ${T2wInputImages} | sed 's/@/ /g' | sed -e  's/^[ \t]*//'`
```
This suggests that someone's setup was passing whitespace or an extra `@` on the front of the argument value.  Since the loop over images uses an unquoted variable, extra whitespace would get ignored anyway, so I'm not sure what the purpose of this is.

TakJim commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +285,19 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
+# NHP options
+opts_AddOptional '--t2wtype' 'T2wType' 'string' "T2w or FLAIR" "T2w"
+opts_AddOptional '--species' 'SPECIES' 'string' "Species (default: Human)" "Human"
+opts_AddOptional '--runmode' 'RunMode' 'int' "Run mode (default: 1)" "1"
```
@takuya-hayashi 
Could you think of candidate mode names, taking tICA as a reference?

takuya-hayashi replied:
1: SetTemplateGradientNonlinearityAverage
2: ACPCAlignment
3: BrainExtraction
4: T2wToT1wRegAndBiasCorrection
5: AtlasRegistration

TakJim commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -99,43 +100,37 @@ echo "PWD = `pwd`" >> $WD/xfms/log.txt
 ${FSLDIR}/bin/invwarp -w ${OutputTransform} -o ${OutputInvTransform} -r ${Reference2mm}
 
 # T1w set of warped outputs (brain/whole-head + restored/orig)
-verbose_echo " --> Generarting T1w set of warped outputs"
```
Sorry, I accidentally deleted the log message. I fixed it in 10b02191571bca3999c1209457504aa6702e1ee9

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -443,6 +540,22 @@ if [ "$CustomBrain" = "NONE" ] ; then
       else
         log_Msg "Processing Modality: $TXw"
       fi
+    TXwFolder=${T2wFolder}
```
The above if block seems to have more indentation for some reason.  It may have been 4-space convention, and here it jumps to 2-space convention with reference to the outer indentation level.

TakJim replied:
There was a mix of 2-space and 4-space indentation. I fixed it in ce7ab5b4dd81aa574f9ad4630f1a280625686087

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -113,6 +113,10 @@ get_batch_options() {
 				index=$(( index + 1 ))
 				;;
+            --Species=*)
+                command_line_specified_subj=${argument#*=}
```
I've deleted --species option in https://github.com/Washington-University/HCPpipelines/pull/350/commits/a17998e71c39d9943e681b019323a3cff5a89293

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -512,7 +516,8 @@ main()
 			--avgrdcmethod="$AvgrdcSTRING" \
-			--topupconfig="$TopupConfig"
+			--topupconfig="$TopupConfig" \
+			--species="$Species"
```
I've deleted --species option in https://github.com/Washington-University/HCPpipelines/pull/350/commits/a17998e71c39d9943e681b019323a3cff5a89293

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +495,513 @@ fi
+            TXwFolder=${T2wFolder}
+            TXwImage=${T2wImage}
+            TXwTemplate=${T2wTemplate}
+            #TXwTemplateBrain=${T2wTemplateBrain}
```
Why is this commented out?

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +495,513 @@ fi
+            ${RUN} ${FSLDIR}/bin/imcp ${TemplateMask} ${AtlasSpaceFolder}/TemplateMask
+            Contrast=T1w
+            TXwTemplate=${AtlasSpaceFolder}/T1wTemplate	
+            #TXwTemplateBrain=${T1wTemplateBrain}
```
Why is this commented out?

glasserm commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -395,12 +410,54 @@ for TXw in $Modalities ; do
-            ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/Magnitude_brain -r ${WD}/Magnitude_brain -w ${WD}/FieldMap_Warp${TXw}.nii.gz -o ${WD}/Magnitude_brain_warpped${TXw}
-            ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_brain_warpped${TXw} -ref ${TXwImageBrain} -out ${WD}/Magnitude_brain_warpped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
-            ;;
+          if [[ ! $SPECIES =~ Marmoset ]] ; then
```
When is this conditional run?

glasserm commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -462,7 +537,7 @@ else
   # QA image
   verbose_echo "      ... Creating QA image"
-  ${FSLDIR}/bin/fslmaths ${WD}/T2w2T1w/T2w_reg -mul ${T1wImage} -sqrt ${WD}/T2w2T1w/sqrtT1wbyT2w -odt float
+  ${FSLDIR}/bin/fslmaths ${WD}/T2w2T1w/T2w_reg -mul ${WD}/${T1wImageBasename}_restore -sqrt ${WD}/T2w2T1w/sqrtT1wbyT2w -odt float
```
Does this need to be NHP conditional?

TakJim replied:
I've add ! Human Conditional in 05e6880686b1e7c3d899febcf3432e4c93f0f085.

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -136,16 +185,71 @@ if [ ! $GradientDistortionCoeffs = "NONE" ] ; then
+  ${GlobalScripts}/CorrectVolumeOrientation --in=${WD}/${vol}.nii.gz --out=${WD}/${vol} --tposition=$TruePatientPosition --sposition=$ScannerPatientPosition $initmat
+    done
+
+  else
```
This is something new for human, but it is probably okay as we have previously recommended people do this outside of the pipeline.

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -217,16 +402,71 @@ elif [[ $UnwarpDir = [yj] || $UnwarpDir = [yj]- || $UnwarpDir = -[yj] ]] ; then
     ShiftTwo="y"
     i=`echo "$i + 1" | bc`
   done
+  if [ $Phase2ndDir = TRUE ] ; then
```
Need to indent 2 spaces.

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -1,8 +1,9 @@
 #!/bin/bash 
+set -e
```
@coalsont do we want this in the scripts?

glasserm replied:
Please remove.

TakJim replied:
I've removed it in https://github.com/Washington-University/HCPpipelines/commit/031e15e2385f9141904472c8bab4e64efcb6a247.

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -67,6 +68,8 @@ opts_AddOptional '--ref2mmmask' 'Reference2mmMask' 'mask' 'reference 2mm brain m
 opts_AddOptional '--fnirtconfig' 'FNIRTConfig' 'file' 'FNIRT configuration file' "${HCPPIPEDIR_Config}/T1_2_MNI152_2mm.cnf"
 
+opts_AddOptional '--brainextract' 'BrainExtract' 'string' 'brain extraction method' ""
+
```
What is this flag for?

glasserm replied:
@takuya-hayashi to determine what happened here.

takuya-hayashi replied:
@TakJim please revert it

glasserm commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -1,8 +1,9 @@
 #!/bin/bash 
+set -e
```
Don't add.

glasserm commented on PreFreeSurfer/scripts/BrainExtraction_FNIRTbased.sh:
```diff
@@ -1,4 +1,9 @@
 #!/bin/bash 
+set -e
```
Remove.

glasserm commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -443,13 +500,31 @@ else
   # Main registration: between corrected T2w and corrected T1w
   verbose_echo "      ... Corrected T2w to T1w"
-  ${FSLDIR}/bin/epi_reg --epi=${WD}/${T2wImageBrainBasename} --t1=${WD}/${T1wImageBasename} --t1brain=${WD}/${T1wImageBrainBasename} --out=${WD}/T2w2T1w/T2w_reg
+  ${FSLDIR}/bin/epi_reg --epi=${WD}/${T2wImageBasename}_restore_brain --t1=${WD}/${T1wImageBasename}_restore --t1brain=${WD}/${T1wImageBasename}_restore_brain --out=${WD}/T2w2T1w/T2w_reg  
```
! Human conditional

TakJim replied:
I've add ! Human Conditional in 05e6880686b1e7c3d899febcf3432e4c93f0f085.

glasserm commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -322,16 +332,21 @@ case $DistortionCorrection in
+            --phasezero=${SpinEchoPhaseEncodeZero}  \
             --scoutin=${ScoutInputName} \
-            --echospacing=${SEEchoSpacing} \
+            --seechospacing=${SEEchoSpacing} \
```
Needs matching change in TopupPreprocessingAll and also keep the current name as a legacy flag.  This also affects fMRIVolume, DiffusionPreprocessing.

coalsont replied:
Legacy flags work like this:

https://github.com/Washington-University/HCPpipelines/blob/d9943356af5ef3282a04ae0f72500d8ec64aae03/PreFreeSurfer/PreFreeSurferPipeline.sh#L187

`--session` is the new flag, and displayed in the help info, the `--subject` flag on the end also works to set the same parameter, but is not displayed in the help info.

If it is an optional parameter, you must specify a default value (you can use "" to get the same behavior as not having specified a default), so that the number of arguments to `opts_AddOptional` works out.

TakJim replied:
Standardized the option name to SEEchoSpacing for consistency across scripts.
The previous --echospacing flag is now treated as a legacy option for backward compatibility.
1100174f0e43b327f58dccad5cd37933ec0f1af5

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -269,7 +528,26 @@ elif [[ $UnwarpDir = [xyij]- || $UnwarpDir = -[xyij] ]] ; then
+   ${GlobalScripts}/bet4animal ${WD}/SBRef ${WD}/SBRef_brain -m -z $species -c $centerx $centery $centerz -f 0.4
+   ${GlobalScripts}/bet4animal ${WD}/PhaseOne_gdc_one ${WD}/PhaseOne_gdc_one_brain -m -z $species -c $centerx $centery $centerz -f 0.4
+  fi
+  ${FSLDIR}/bin/flirt -dof 6 -interp spline -in ${WD}/SBRef_brain -ref ${WD}/PhaseOne_gdc_one_brain -omat ${WD}/SBRef2PhaseOne_gdc.mat -out ${WD}/SBRef2PhaseOne_gdc -nosearch
```
! Human Conditional

TakJim replied:
I've add ! Human conditional in 93a6fd9daf82d21c8df4912df025c3aaef501d7d.

glasserm commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -290,14 +568,32 @@ VolumeNumber=$((0 + 1))
 ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/PhaseOne_gdc -r ${WD}/Mask --premat=${WD}/MotionMatrix_${vnum}.mat -w ${WD}/WarpField_${vnum} -o ${WD}/PhaseOne_gdc_dc
 ${FSLDIR}/bin/fslmaths ${WD}/PhaseOne_gdc_dc -mul ${WD}/Jacobian_${vnum} ${WD}/PhaseOne_gdc_dc_jac
 
+if [ $Phase2ndDir = TRUE ] ; then
```
No indent, please use two spaces.

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,263 @@
+  echo $Subject
+
+  #Scan settings
+  source ${StudyFolder}/${Subject}/RawData/hcppipe_conf.txt
```
We should probably add a comment explaining what variables this defines for us, so other users can either define them manually or set up similar files.

We could also consider making it conditional on whether the file exists and having lines to set them manually in this batch script.

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+# All rights reserved.
+
+## Species specific variables
+if [[ "$SPECIES" == *Human* ]] ; then
```
With the change to argument-passing in PreFSBatchNHP, this should either be using `$1`, or do `SPECIES=$1` near the top.  At present, it is still looking at the value assigned to `$SPECIES` in the sourcing code, so a bare `source SetUpSPECIES.sh Macaque 0.35` without setting `$SPECIES` in the batch script would fail.  Similar for `$2` versus `$StrucRes`.

glasserm replied:
${UnwarpDir} needs to be passed in as an argument.

glasserm replied:
Also MION

coalsont replied:
If UnwarpDir isn't used to set filenames, etc, then this file shouldn't need to know about it (and I think it shouldn't be responsible for setting it, just like it doesn't set the input filenames).

Looks like MION is only used to change the fix training data, and currently uses Macaque MION training data regardless of species (I would guess training doesn't exist for other species).  Again, this is specific to the input data, not necessarily global to the species, so I would consider not having the species setup file interact with it.  Instead, the batch file can have a section that says something like:

```
#if you used MION, change the fix training data by uncommenting this line and providing the path to the training file if necessary.
#TrainingData=NHPHCP_Macaque.USPIO
```

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+	BiasFieldSmoothingSigma="5.0"
+	T1wTemplate="${HCPPIPEDIR_Templates}/MNI152_T1_${StrucRes}mm.nii.gz"  
+	T1wTemplateBrain="${HCPPIPEDIR_Templates}/MNI152_T1_${StrucRes}mm_brain.nii.gz" 
+	T1wTemplate2mm="${HCPPIPEDIR_Templates}/MNI152_T1_2mm"  
```
Having "2mm" in the variable name doesn't seem like a great choice, though it is our fault for putting it in the argument name.  Consider `T1wTemplateGrayord`?

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+	T2wTemplateBrain="${HCPPIPEDIR_Templates}/ChimpYerkes29_T2w_${StrucRes}mm_brain.nii.gz" 
+	T2wTemplate2mm="${HCPPIPEDIR_Templates}/ChimpYerkes29_T2w_1.6mm.nii.gz" 
+	TemplateMask="${HCPPIPEDIR_Templates}/ChimpYerkes29_T1w_${StrucRes}mm_brain_mask.nii.gz" 
+	Template2mmMask="${HCPPIPEDIR_Templates}/ChimpYerkes29_T1w_1.6mm_brain_mask.nii.gz" 
```
Unlike the other NHP species, these aren't in an `NHP_NNP` subfolder.

TakJim replied:
It was decided in the meeting that @takuya-hayashi will organize and share the NHP_NNP template directory.

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+	betradius="35"
+	betbiasfieldcor="TRUE"
+	betfraction="0.2"
+	bettop2center="30"           # distance in mm from the top of FOV to the center of brain in robustroi
```
Indent the conditional code.

TakJim replied:
I've fixed it in 704ca3a7c2c540823d90b045c33eec220e92b2fe.

coalsont commented on PreFreeSurfer/scripts/AnatomicalAverage.sh:
```diff
@@ -36,6 +36,8 @@ opts_AddOptional '--working-dir' 'wdir' 'dir' "temporary working directory"
 
 opts_AddOptional '--cleanup' 'cleanup' 'yes OR no' "whether to delete the working directory (if set via --working-dir), default yes" "yes"
 
+opts_AddOptional '--verbose' 'verbose' 'yes OR no' "verbose output" "no"
```
This doesn't appear to be used.

TakJim replied:
I've removed it in c3ccf76fcfd66074cb81403df0f451070d52cfd6.

coalsont commented on PreFreeSurfer/scripts/AnatomicalAverage.sh:
```diff
@@ -51,12 +53,14 @@ log_Check_Env_Var FSLDIR
 
 if ((${#imagelist[@]} < 2)) ; then
     log_Err_Abort "Must specify at least two images to average"
+
```
Try not to add blank lines in simple blocks like this.  Extra lines can help visually separate logical sections of a script, but otherwise they usually don't help.

TakJim replied:
sorry for the unnecessary blank lines. removed it.
 05f48a9c6ad4dcb6f9abb093800e0e84a92cd45e

coalsont commented on PreFreeSurfer/scripts/BrainExtraction_FNIRTbased.sh:
```diff
@@ -1,4 +1,9 @@
+#  environment: as in SetUpHCPPipeline.sh   (or individually: FSLDIR, HCPPIPEDIR_Templates)
 
 # Requirements for this script
 #  installed versions of: FSL
```
We don't need two of these sections, revert.

coalsont commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -395,12 +410,54 @@ for TXw in $Modalities ; do
+            ${FSLDIR}/bin/convert_xfm -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -concat ${WD}/Fieldmap2${TXwImageBasename}_tmp.mat  ${WD}/Fieldmap2${TXwImageBasename}_init.mat
+            rm -f ${WD}/Fieldmap2${TXwImageBasename}_tmp.mat                        
+          fi
             ;;
```
Please make indentation locally consistent (see also the commands before the conditional).  The outer `for` and the `case` were both 4-space convention, though the internal conditionals somehow ended up with 2 additional spaces instead of being consistent (and the other blocks in the file are 2-space...).

TakJim replied:
I've fixed it in dc8a2808218b39652253eeb8de6db81f6c8687db.

coalsont commented on global/scripts/CorrectVolumeOrientation:
```diff
@@ -0,0 +1,232 @@
+#! /bin/bash
```
We have used ".sh" as the file extension for all our other pipelines scripts (there are some exceptions in global/scripts, but it would probably be better if they also had `.sh`), please add it.

TakJim replied:
Thanks for pointing that out.
I’ve added the .sh extension to the script and updated all corresponding calls accordingly.
8e4985d33312b4b8cce4845597d1ba5e2ae33992

coalsont commented on global/scripts/CorrectVolumeOrientation:
```diff
@@ -0,0 +1,232 @@
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
We don't have copyright notices inside our other scripts.  There is a `LICENSE.md` file in the root of the repository instead (BSD 3-clause, though we should maybe update the copyright year).

TakJim replied:
I've deleted inline copyright in 26c54cfed961e1d08c8db4129faf04fae201b1d5.

coalsont commented on global/scripts/CorrectVolumeOrientation:
```diff
@@ -0,0 +1,232 @@
+
+$ CorrectVolumeOrientation --in=MRI.nii.gz --tposition=HFSx --sposition=HFP --out=MRI2reorient.nii.gz
+
+Ex) #2. structural MRI scanned in true position of HFS on day1 (T1w_d1_HFS) and fMRI scanned in true posiiton in HFSx on day2 (fMRI_d2_HFSx) and additional fast T1w scanned for registration purpose on day2 (T1w_d2_HFSx), where scanner's patient orientaion is set to HFS on day1 and HFP on day2. In this case, the fast T1w scanned on day2 is corrected for the head orientation and registrated to the  ...
```
It would probably be a good idea to add manual word wrapping to this entire description, as `newops.shlib` doesn't apply word wrapping code to it (only the argument descriptions get word wrapped, and that code is specialized to indent wrapping to avoid confusion with the next argument).

coalsont commented on global/scripts/bet4animal:
```diff
@@ -0,0 +1,324 @@
+#!/bin/bash
```
If this script doesn't need to be the same exact name as something else, I suggest adding the `.sh` extension.

takuya-hayashi replied:
Thanks. I agree it is reasonable to add the extension. However, I just followed the original 'bet', which is also bash script. I think we agreed to remove this script from the HCP pipeline since it is already bundled in FSL. -> @TakJim

glasserm replied:
6.0.6

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -326,6 +345,29 @@ then
 
+# Convert the --runmode string argument into a numeric code
+case "$RunMode" in
+    Default)
```
Allow empty string as another way to get default behavior
```suggestion
    Default|"")
```

coalsont commented on PreFreeSurfer/scripts/AnatomicalAverage.sh:
```diff
@@ -36,6 +36,8 @@ opts_AddOptional '--working-dir' 'wdir' 'dir' "temporary working directory"
 
```
This script has been changed to non-executable by the PR, please `chmod +x` it and commit again.

coalsont replied:
Several other scripts in this PR have also been changed to non-executable.

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,263 @@
+SPECIES=""
+RunMode=""
+BrainExtract=""
+StrucRes=""
```
The way we do other Batch.sh files, we would tell users to edit the file and put the values here.  The argument parsing in other Batch.sh files is largely an afterthought (I think it streamlined someone's testing process), the instructions say things like "copy the Batch.sh and edit the section that says (whatever) to set the StudyFolder, Subjlist, (etc), then run it without arguments".

TakJim replied:
I removed these initialization in 285c97d1b2d58ac417d0ef479c78b721e893dd34.

TakJim commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -395,12 +410,54 @@ for TXw in $Modalities ; do
         ${FIELDMAP_METHOD_OPT} | ${SIEMENS_METHOD_OPT} | ${GE_HEALTHCARE_LEGACY_METHOD_OPT} | ${GE_HEALTHCARE_METHOD_OPT} | ${PHILIPS_METHOD_OPT})
             ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/Magnitude -r ${WD}/Magnitude -w ${WD}/FieldMap_Warp${TXw}.nii.gz -o ${WD}/Magnitude_warpped${TXw}
             ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_warpped${TXw} -ref ${TXwImage} -out ${WD}/Magnitude_warpped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
+          if [[ ! $SPECIES =~ Marmoset ]] ; then
```
I've fixed it in f9791caa68627d869555c203b446122ed1a590dc.

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,263 @@
+
+# Set default values for optional parameters
+BrainExtract=${BrainExtract:-INVIVO}
+RunMode=${RunMode:-Default}
```
I've fixed it 870e2d468e618c47db54e9fe9df74f081e00d2db.

TakJim commented on PreFreeSurfer/scripts/AnatomicalAverage.sh:
```diff
@@ -99,16 +105,17 @@ im1=`echo $newimlist | awk '{ print $1 }'`;
+
     else
-		cp $FSLDIR/etc/flirtsch/ident.mat ${im1}_to_im1_linmask.mat
+  cp $FSLDIR/etc/flirtsch/ident.mat ${im1}_to_im1_linmask.mat
```
I've fixed it in ad2e9a080c734dc0666018b787041bc65168bfaa.

TakJim commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +495,513 @@ fi
+        ${RUN} ${FSLDIR}/bin/imcp ${T2wTemplate} ${AtlasSpaceFolder}/T2wTemplate
+        ${RUN} ${FSLDIR}/bin/imcp ${T2wTemplateBrain} ${AtlasSpaceFolder}/T2wTemplateBrain
+        TXwTemplate=${AtlasSpaceFolder}/T2wTemplate	
+        #TXwTemplateBrain=${AtlasSpaceFolder}/T2wTemplateBrain
```
TXwTemplateBrain was an unused parameter, so I removed it.
c3e90a2b08b6384ae8e9fd8f573c6141b628d758

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,305 @@
+
+# species specific config
+# If StrucRes is not provided, SetUpSPECIES.sh will use species-specific default
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh "$SPECIES" "$StrucRes"
```
Also we should use the standard option parsing for this script like all the other ones.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+    CorticalScaleFactor="1"
+
+    #PreFreeSurferPipeLineBatch.sh
+    BrainExtract="${BrainExtract:-INVIVO}"
```
This construct is opaque to me.  What are these lines doing?

TakJim replied:
These initializations of BrainExtract were unnecessary, so I removed them.
51f38d3183d9872e1f9416a10894150eb03320d0

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +496,513 @@ fi
+            TXwFolder=${T1wFolder}
+            TXwImage=${T1wImage}
+            TXwTemplate=${T1wTemplate}
+            # Create reference volumes if the resolution of raw image differs from TXwTemplate - TH Mar 2023 
```
Not addressed yet.

glasserm commented on PreFreeSurfer/scripts/BrainExtraction_FNIRTbased.sh:
```diff
@@ -99,12 +130,62 @@ echo " " >> $WD/log.txt
-${FSLDIR}/bin/flirt -interp spline -dof 12 -in "$Input" -ref "$Reference" -omat "$WD"/roughlin.mat -out "$WD"/"$BaseName"_to_MNI_roughlin.nii.gz -nosearch
-verbose_echo " ... non-linear registration to 2mm reference"
-${FSLDIR}/bin/fnirt --in="$Input" --ref="$Reference2mm" --aff="$WD"/roughlin.mat --refmask="$Reference2mmMask" --fout="$WD"/str2standard.nii.gz --jout="$WD"/NonlinearRegJacobians.nii.gz --refout="$WD"/IntensityModulatedT1.nii.gz --iout="$WD"/"$BaseName"_to_MNI_nonlin.nii.gz --logout="$WD"/NonlinearReg.txt --intout="$WD"/NonlinearIntensities.nii.gz --cout="$WD"/NonlinearReg.nii.gz --config="$FNIRT ...
+if [ "$BrainExtract" = INVIVO ] ; then
```
Since all these options are NHP specific, I wonder if it would be better code organization and usability to do a if not human conditional around this stuff since we don't really know if any of those additional options would actually work for humans (e.g., EXVIVO).

glasserm commented on global/scripts/bet4animal:
```diff
@@ -0,0 +1,324 @@
+echo "                  4: Night monkey (xscale = 4.0,  yscale = 3.0,  zscale = 5.0,  radius = 20)"
+echo "                  5: Rat          (xscale = 8.0,  yscale = 6.0,  zscale = 10.0, radius = 10)"
+echo "                  6: Mouse        (xscale = 16.0, yscale = 12.0, zscale = 20.0, radius =  5)"
+echo "                  7: Mini pig     (xscale = 3.0,  yscale = 2.6,  zscale = 2.6,  radius = 25)"
```
Do we need to depend on a specific FSL version?

TakJim commented on PreFreeSurfer/scripts/BiasFieldCorrection_T1wOnly.sh:
```diff
@@ -38,6 +38,8 @@ opts_AddOptional '--oT1brain' 'OutputT1wRestoredBrainImage' ' ' "output correcte
 
 opts_AddOptional '--bfsigma' 'BiasFieldSmoothingSigma' 'value' "Bias field smoothing Sigma (Default 20)" "20"
 
+opts_AddOptional '--strongbias' 'StrongBias' 'flag' "use stronger bias field correction" "FALSE"
```
Added missing initialization for StrongBiasFlag.
84ea88a588a74cc53c2aeb9c182a21bcf9787fe6

TakJim commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -84,37 +104,86 @@ echo " " >> $WD/log.txt
+  ${FSLDIR}/bin/flirt -interp spline -in "$WD"/robustroi.nii.gz -ref "$WD"/Reference -init "$WD"/roi2std_init.mat -omat "$WD"/roi2std.mat -out "$WD"/acpc_final.nii.gz -nosearch -dof 12 # -inweight "$WD"/robustroi_brain_mask -refweight "$Reference2mmMask" - 0609 did not work with these inweight & refweight
+else              # NONE 
+  verbose_echo " --> Registering cropped image to MNI152 (12 DOF)"
+  ${FSLDIR}/bin/flirt -interp spline -in "$WD"/robustroi.nii.gz -ref "$Reference2mm" -omat "$WD"/roi2std.mat -out "$WD"/acpc_final.nii.gz -searchrx -30 30 -searchry -30 30 -searchrz -30 30
```
Thanks for pointing that out! I had mistakenly modified the Human process section — I’ve reverted it back to the original behavior.
e6b5554607273a67cc1e9b85bfd8fee042880472

TakJim commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,471 @@
+	bettop2center="60"           # Distance between top of FOV and center of brain
+	betspecieslabel="1"          # bet4animal species label
+
+	StrucRes=${StrucRes:-0.8}
```
I agree, making StrucRes mandatory in the Batch makes sense. I’ll update the script accordingly.

TakJim commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +498,499 @@ fi
+        --ot2rest=${AtlasSpaceFolder}/${T2wImage}_restore \
+        --ot2restbrain=${AtlasSpaceFolder}/${T2wImage}_restore_brain \
+        --fnirtconfig=${FNIRTConfig} \
+        --brainextract=${BrainExtract}
```
Sorry, I've removed it in .
4407dac16dc1886db185fc2973ab008f9dbd82b2

coalsont commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -120,61 +149,205 @@ fi
+        Phase2ndSet=""
+    fi
+
+    if [[ ("$TruePatientPosition" = "HFSx" || "$TruePatientPosition" = "FFSx" || "$TruePatientPosition" = "HFS" || "$TruePatientPosition" = "FFS" ) && ( "$TruePatientPosition" != "$ScannerPatientPosition") ]] ; then
```
Why is this conditional this complicated?  Should we error on non-recognized true position string instead of treating it as "trust the scanner"?

coalsont commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -83,38 +104,87 @@ echo "date: `date`" >> $WD/log.txt
 verbose_echo " --> Geting a 6 DOF approximation"
-$CARET7DIR/wb_command -convert-affine -from-flirt "$WD"/full2std.mat "$Input".nii.gz "$Reference" -to-world "$WD"/full2std_world.mat
+#${FSLDIR}/bin/aff2rigid "$WD"/full2std.mat "$OutputMatrix"
+${CARET7DIR}/wb_command -convert-affine -from-flirt "$WD"/full2std.mat "$Input".nii.gz "$Reference".nii.gz -to-world "$WD"/full2std_world.mat
```
Change all of these back to not add a `.nii.gz` onto `"$Reference"`, since we are passing these as full filenames again.

coalsont commented on PreFreeSurfer/scripts/BiasFieldCorrection_T1wOnly.sh:
```diff
@@ -78,11 +81,27 @@ echo "PWD = `pwd`" >> $WDir/log.txt
+else
+	StrongBias=NONE
+	StrongBiasFlag=""
+fi
```
Typo in conditional, and unused de-conversion of `0` from false-like values to `NONE`.
```suggestion
# parse/sanity check for TRUE or FALSE, yes/no, etc
StrongBias=$(opts_StringToBool "$StrongBias")

# if TRUE, pass --strongbias to fsl_anat
if ((StrongBias)); then
	StrongBiasFlag="--strongbias"
else
	StrongBiasFlag=""
fi
```

coalsont commented on PreFreeSurfer/scripts/BiasFieldCorrection_T1wOnly.sh:
```diff
@@ -78,11 +81,27 @@ echo "PWD = `pwd`" >> $WDir/log.txt
 # Compute T1w Bias Normalization using fsl_anat function
 
-${FSLDIR}/bin/fsl_anat -i $T1wImage -o $WD --noreorient --clobber --nocrop --noreg --nononlinreg --noseg --nosubcortseg -s ${BiasFieldSmoothingSigma} --nocleanup
+if [ $(${FSLDIR}/bin/imtest $T1wBrain) = 0 ] ; then
```
What happens here if someone reruns human PreFS without deleting?

TakJim commented on PreFreeSurfer/scripts/BrainExtraction_FNIRTbased.sh:
```diff
@@ -99,12 +130,62 @@ echo " " >> $WD/log.txt
+    fslmaths "$Reference2mm" -mas "$Reference2mmMask" "$WD"/ReferenceBrain
+    fslmaths "$Input" "$WD"/"$BaseName"
+    ${HCPPIPEDIR_Global}/bet4animal "$WD"/"$BaseName" "$WD"/"$BaseName"_brain_initI $BetOpts     # "-c 48 56 51 -r 36" for A21051401
+    if [ ! $SPECIES = Human ] ; then
```
@coalsont 
I've add SPECIES option in 764eb822fc853e115e2de170ab537715f0c61f55.

@takuya-hayashi 
we’ve removed bet4animal because it’s now part of FSL. However, it is still being called in this part of the script.
Could you please advise how we should modify this section accordingly?

takuya-hayashi replied:
please change from 

${HCPPIPEDIR_Global}/bet4animal

to 

${FSLDIR}/bin/bet4animal

takuya-hayashi commented:
Please change from
${HCPPIPEDIR_Global}/bet4animal
to
${FSLDIR}/bin/bet4animal


> Message ID: <Washington-University/HCPpipelines/pull/350/review/3369413198
> @github.com>
>

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -125,6 +127,7 @@ if [ ! "${T2wImage}" = "NONE" ] ; then
   ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${T2wRestore} -r ${Reference} -w ${OutputTransform} -o ${OutputT2wImageRestore}
   ${FSLDIR}/bin/applywarp --rel --interp=nn -i ${T2wRestoreBrain} -r ${Reference} -w ${OutputTransform} -o ${OutputT2wImageRestoreBrain}
   ${FSLDIR}/bin/fslmaths ${OutputT2wImageRestore} -mas ${OutputT2wImageRestoreBrain} ${OutputT2wImageRestoreBrain}
+  ${FSLDIR}/bin/imcp ${OutputT2wImage} ${OutputT2wImage}_orig
```
What is this for?

glasserm replied:
Remove.

glasserm commented on Examples/Scripts/SetUpHCPPipeline.sh:
Don't change this one.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,457 @@
+SPECIES="$1"
+StrucRes="$2"
+UnwarpDir="$3"
+MION="$4"
```
Takuya requests USPIO instead of MION.

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+# If StrucRes is not provided, SetUpSPECIES.sh will use species-specific default
+#if you used MION, change the fix training data by uncommenting this line and providing the path to the training file if necessary.
+#TrainingData=NHPHCP_Macaque.USPIO
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh "$SPECIES" "$StrucRes" "$UnwarpDir"
```
Make commented out variable names for the user to overide defaults from Set species.

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+    else
+        echo "  WARNING: ${StudyFolder}/${Subject}/RawData/hcppipe_conf.txt not found. Please set values manually."
+
+        ## Structural MRI
```
Remove the things not related to PreFreeSurfer.

glasserm commented on PreFreeSurfer/scripts/AtlasRegistrationToMNI152_FLIRTandFNIRT.sh:
```diff
@@ -117,6 +118,7 @@ ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${T1wImage} -r ${Reference} -w
 ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${T1wRestore} -r ${Reference} -w ${OutputTransform} -o ${OutputT1wImageRestore}
 ${FSLDIR}/bin/applywarp --rel --interp=nn -i ${T1wRestoreBrain} -r ${Reference} -w ${OutputTransform} -o ${OutputT1wImageRestoreBrain}
 ${FSLDIR}/bin/fslmaths ${OutputT1wImageRestore} -mas ${OutputT1wImageRestoreBrain} ${OutputT1wImageRestoreBrain}
+${FSLDIR}/bin/imcp ${OutputT1wImage} ${OutputT1wImage}_orig
```
Remove.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -326,6 +347,29 @@ then
+        ;;
+    *)
+        echo "Error: invalid runmode '$RunMode'"
+        exit 1
```
```suggestion
        log_Err_Abort "Error: invalid runmode '$RunMode'"
```

Also, below this block, add either an `if [[ "$BrainExtract" != INVIVO && "$BrainExtract" != EXVIVO ]]` or a `case "$BrainExtract" in` to make the script error if the user has done something silly like `--brainextract=UNKNOWN`.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +280,22 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
+opts_AddOptional '--betfraction' 'betfraction' 'float' "Fraction for BET (default: 0.3)" "0.3"
+opts_AddOptional '--bettop2center' 'bettop2center' 'int' "Distance from top to center for BET (default: 86)" "86"
+opts_AddOptional '--brainextract' 'BrainExtract' 'string' "Brain extraction method (default: INVIVO)" "INVIVO"
+opts_AddOptional '--betspecieslabel' 'betspecieslabel' 'string' "default: " ""
```
This should have a default of 0, to match the `--species` default of "Human".  We should probably list the current accepted species and/or say "refer to the help info of bet4animal for the full list of codes for supported species".

Having both `--species` and `--betspecieslabel` arguments seems redundant, but would allow users to use species that FSL adds to bet4animal without needing a matching update to the pipelines.  However, as-is, that flexibility prevents us from sanity checking to catch if the user specified something like `--species="Homo Sapiens"` or `--species=HUMAN`.

It appears that the only things inside the pipelines (which doesn't count `SetUpSPECIES.sh`, but that already has a sanity check that will error on anything unrecognized) that our code tests `$SPECIES` against are `"Human"` and `*Marmoset*`.  So, it would be largely future-proof to remove `--species` entirely from the pipelines and convert them all to testing if `betspecieslabel` is 0 or not (possibly using an `--is-human` yes/no argument for most of them), and pass `betspecieslabel` into T2wToT1wDistortionCorrection... to test it against the appropriate integer to detect marmoset.  I imagine NHP fMRIVolume may also call `bet4animal`, and maybe TransmitBias, since they both take in raw scans (and skull can be a problem for registration).  Consistency would suggest using the same arguments even in pipelines that don't run `bet4animal`, so it might look a little odd that we pass `--betspecieslabel` to, say, fMRISurface (if that needs NHP adjustments).

Yes, this is a substantial change, so probably wait for @glasserm to respond before actually doing it.

coalsont commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -80,9 +96,22 @@ log_Check_Env_Var FSLDIR
+    species=3
+elif [[ "$SPECIES" =~ NightMonkey ]] ; then
+    species=4
+fi
```
Does this deliberately treat everything "beyond" night monkey as "species 4"?  I hope not, because the obvious way to convert these is to have the entire list of supported values, and then error for anything not recognized.  `case` is more elegant though, because you don't need to repeat the variable.

TakJim replied:
Fixed it by using common case statement for convert species to betspecieslabel
e8444c4a9ca3ac5e1de1af652be8c68a2852574b

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,466 @@
+    else
+        CustomTrainingData="NHPHCP_Marmoset"
+        if [ "$MION" = 1 ] ; then 
+            TrainingData=/mnt/pub/PROJ/NHP_NNP/MacaqueRhesus/NHPHCP_Macaque.USPIO.pyfix/NHPHCP_Macaque.USPIO
```
I think we should just have this in the eventual `IcaFixProcessingBatchNHP.sh` instead of testing for `MION` in `SetUpSPECIES.sh` at all:
```
#uncomment this (and provide full path if necessary) if your data used MION/USPIO:
#TrainingData=NHPHCP_Macaque.USPIO
```

However, we can put this off until we actually work on the FIX pipeline and just remove them from this PR.

Side note, it also looks like the unconditional lines shouldn't be `CustomTrainingData=`.

coalsont commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -83,38 +104,87 @@ echo "date: `date`" >> $WD/log.txt
+if [[ "$SPECIES" != "Human" ]] ; then
+  fslmaths "$Reference2mm" "$WD"/Reference
+  fslmaths "$Reference2mm" -mas "$Reference2mmMask" "$WD"/ReferenceBrain
+fi
```
Give a more useful error than "file not found" (if someone makes a bad edit to the pipelines).  Also, document this requirement in the `--ref2mm` and `--ref2mmmask` descriptions.
```suggestion
if [[ "$SPECIES" != "Human" ]] ; then
  if [[ "$Reference2mm" == "" || "$Reference2mmMask" == "" ]]
  then
    log_Err_Abort "--ref2mm and --ref2mmmask are required when species is not 'Human'"
  fi
  fslmaths "$Reference2mm" "$WD"/Reference
  fslmaths "$Reference2mm" -mas "$Reference2mmMask" "$WD"/ReferenceBrain
fi
```

TakJim commented on PreFreeSurfer/scripts/BrainExtraction_FNIRTbased.sh:
```diff
@@ -1,7 +1,7 @@
 
 # Requirements for this script
-#  installed versions of: FSL
+#  installed versions of: FSL=6.0.6
```
Thank you for the suggestion! I’ve updated the requirement to FSL>=6.0.6 to clarify that later versions are supported as well.
4a92ef0bc69dce5af980462dc176fafaeaa73d72

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,363 @@
+#! /bin/bash
```
I will also note that `set -eu` and `debug.shlib` will carry over to anything in the Batch.sh after `SetUpSPECIES.sh`, but using an undefined variable in a Batch.sh is probably also unintended, so...

coalsont replied:
Also note that using `newopts.shlib` will cause the Batch.sh to print the hostname and the arguments to `SetUpSPECIES.sh`.

TakJim replied:
Thank you for your suggestion! I fixed it.
7026d9ccf851364481ce9f33d0e61fc4eb50cd6e

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,363 @@
+# All rights reserved.
+
+source "$HCPPIPEDIR/global/scripts/newopts.shlib" "$@"
+source "$HCPPIPEDIR/global/scripts/debug.shlib" "$@"         # Debugging functions; also sources log.shlib
```
Don't imply that logging isn't sourced by newopts.
```suggestion
source "$HCPPIPEDIR/global/scripts/newopts.shlib" "$@"     #also sources log.shlib
source "$HCPPIPEDIR/global/scripts/debug.shlib" "$@"
```

TakJim replied:
Thank you for the suggested fix!
a1fb694c04a20c39b2beeba12ba42d8976b4f47e

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+    # If the config file is not found, please set values manually.
+    # ------------------------------------------------------------------------------
+    if [ -f "${StudyFolder}/${Subject}/RawData/hcppipe_conf.txt" ]; then
+        source "${StudyFolder}/${Subject}/RawData/hcppipe_conf.txt"
```
Added the example
437ea25cb8173df79267bb7217f8019dc1bddcda

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+SPECIES="Human"
+RunMode="Default"
+BrainExtract="INVIVO"
+StrucRes="0.8"
```
I added the default UnwarpDir.
I didn’t add anything for MION since it’s not being used.
454094c358c959be6bbc44ff7c6dde948f442ae5

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,329 @@
+# If StrucRes is not provided, SetUpSPECIES.sh will use species-specific default
+#if you used MION, change the fix training data by uncommenting this line and providing the path to the training file if necessary.
+#TrainingData=NHPHCP_Macaque.USPIO
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh --species="$SPECIES" --structres="$StrucRes" 
```
Thank you for the suggested fix!


0c9582fa41b65ac2d5d3121751b23e336fb6e095

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,308 @@
+        # optional variables of sMRI
+        #StrucTruePatientPosition=HFS                                          # HFS: head-first-supine (default) or HFSx: head-first sphinx or FFSx: foot-first sphinx
+        #StrucScannerPatientPosition=HFS                                       # scanner's patient position in the DICOM, HFS (default) or HFP or FFS or FFP
+        #T1wSampleSpacing=".00000710000000000000"                              # readout time (optional for B0 distortion correction)
```
state that time is in seconds.

glasserm commented on Examples/Scripts/hcppipe_conf.txt:
```diff
@@ -0,0 +1,14 @@
+# optional variables of sMRI
+StrucTruePatientPosition=HFS                                          # HFS: head-first-supine (default) or HFSx: head-first sphinx or FFSx: foot-first sphinx
+StrucScannerPatientPosition=HFS                                       # scanner's patient position in the DICOM, HFS (default) or HFP or FFS or FFP
+T1wSampleSpacing=".00000710000000000000"                              # readout time (optional for B0 distortion correction)
```
State that time is in seconds.

TakJim replied:
fixed it in 730fb92b71ec3783e1b5d17e75c33a88be622a2a.

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +280,21 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
+opts_AddOptional '--betfraction' 'betfraction' 'float' "Fraction for BET (default: 0.3)" "0.3"
+opts_AddOptional '--bettop2center' 'bettop2center' 'int' "Distance from top to center for BET (default: 86)" "86"
+opts_AddOptional '--brainextract' 'BrainExtract' 'string' "Brain extraction method (default: INVIVO)" "INVIVO"
+opts_AddOptional '--flair' 'flairString' 'TRUE/FALSE' "Indicates whether to add T2-weighted image as a phase zero volume" "FALSE"
```
This option name and explanation is confusing.

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,308 @@
+        Please prepare hcppipe_conf.txt.
+        Alternatively, uncomment the lines in the Batch File, set appropriate values manually, and run again."
+
+        ## Structural MRI (sMRI) - multiple scans can be separated by a space (" ")
```
Every argument in the pipeline call needs to be settable above the pipeline call without editing set species.

TakJim replied:
By uncommenting the parameters in SetSpecies.sh, I made it possible to configure them within the Batch script.
https://github.com/Washington-University/HCPpipelines/pull/350/commits/2f3c4b17c44cbe36efd7a72750c50386ae7521dd

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +501,491 @@ fi
+                    rm -r ${T2wFolder}/T2wToT1wDistortionCorrectAndReg
+                fi
+
+                if [ $(imtest ${T2wFolder}/T2w) = 1 ] && [ "$flair" = "FALSE" && "$SPECIES" != "Human" ] ; then    # added T2w as a phase zero volume - TH Jan 2023
```
$flair is not set.

glasserm replied:
Change to $UsePhaseZero

TakJim replied:
Rename the variable name in cb2883061de9f41177e1778acd18ac96c669d765.

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -326,6 +351,28 @@ then
+        log_Err_Abort "Error: invalid runmode '$RunMode'"
+        ;;
+esac
+
```
Add `opts_StringToBool` on `$UsePhaseZero`:

https://github.com/Washington-University/HCPpipelines/blob/12e8de201e02169d47dea2a908331eed93bd37bf/global/scripts/RSNregression.sh#L72

TakJim commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +280,21 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
 opts_AddOptional '--usejacobian' 'UseJacobian' 'TRUE or FALSE' "Whether to use jacobian modulation when correcting spin echo fieldmaps for gradient distortion" "True" # NOT IN THE ORIGNAL SCRIPT
+
+# NHP options
+opts_AddOptional '--t2wtype' 'T2wType' 'string' "T2w or FLAIR" "T2w"
```
Remove unuse 'T2wType'
https://github.com/Washington-University/HCPpipelines/commit/8c24e008ee4addb2f7e2119a141776c5d48e2455

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +503,491 @@ fi
+                    rm -r ${T2wFolder}/T2wToT1wDistortionCorrectAndReg
+                fi
+
+                if [ $(imtest ${T2wFolder}/T2w) = 1 ] && [ "$UsePhaseZero" = "FALSE" && "$SPECIES" != "Human" ] ; then    # added T2w as a phase zero volume - TH Jan 2023
```
fix logic (`UsePhaseZero` is never "FALSE", it is 0 or 1, and we want the true case, not the false case), reorder to make the "not human" test more obvious, use `[[ ]]` bash grammar instead of `[` builtin executable-like behavior.
```suggestion
                if [[ "$SPECIES" != "Human" && "$(imtest "$T2wFolder"/T2w)" == "1" && "$UsePhaseZero" == "1" ]] ; then    # added T2w as a phase zero volume - TH Jan 2023
```

coalsont replied:
Since the option defaults to false anyway, do we actually want the "not human" test here?  Technically, it means that the option description should say something like "(disabled for human subjects)" to be accurate for the current logic.

coalsont replied:
@glasserm thoughts on whether `--use-t2w-phase-zero=TRUE` should be nonfunctional if someone chooses to add it to a human run?  It defaults to false, so it matches current master behavior by default.

glasserm replied:
That would be a new feature, and should be allowed.

TakJim commented on Examples/Scripts/hcppipe_conf.txt:
```diff
@@ -0,0 +1,14 @@
+StrucTruePatientPosition=HFS                                          # HFS: head-first-supine (default) or HFSx: head-first sphinx or FFSx: foot-first sphinx
+StrucScannerPatientPosition=HFS                                       # scanner's patient position in the DICOM, HFS (default) or HFP or FFS or FFP
+T1wSampleSpacing=".00000710000000000000"                              # readout time (optional for B0 distortion correction)
+T2wSampleSpacing=".00000210000000000000"                              # readout time (optional for B0 distortion correction)
```
fixed it in 730fb92b71ec3783e1b5d17e75c33a88be622a2a.

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,308 @@
+        #StrucTruePatientPosition=HFS                                          # HFS: head-first-supine (default) or HFSx: head-first sphinx or FFSx: foot-first sphinx
+        #StrucScannerPatientPosition=HFS                                       # scanner's patient position in the DICOM, HFS (default) or HFP or FFS or FFP
+        #T1wSampleSpacing=".00000710000000000000"                              # readout time (optional for B0 distortion correction)
+        #T2wSampleSpacing=".00000210000000000000"                              # readout time (optional for B0 distortion correction)
```
fixed it in 730fb92b71ec3783e1b5d17e75c33a88be622a2a.

glasserm commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +280,20 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
+opts_AddOptional '--betfraction' 'betfraction' 'float' "Fraction for BET (default: 0.3)" "0.3"
+opts_AddOptional '--bettop2center' 'bettop2center' 'int' "Distance from top to center for BET (default: 86)" "86"
+opts_AddOptional '--brainextract' 'BrainExtract' 'string' "Brain extraction method (default: INVIVO)" "INVIVO"
+opts_AddOptional '--use-t2w-phase-zero ' 'UsePhaseZero' 'TRUE/FALSE' "Indicates whether to add T2-weighted image as a phase zero volume" "FALSE"
```
"for dark-CSF T2w contrast acquisition types (e.g., FLAIR)"

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,356 @@
+#TemplateMask="${HCPPIPEDIR_Templates}/MNI152_T1_${StrucRes}mm_brain_mask.nii.gz" 
+#Template2mmMask="${HCPPIPEDIR_Templates}/MNI152_T1_2mm_brain_mask_dil.nii.gz" 
+
+#### PostFreeSurferPipeLineBatch.sh
```
As you suggested, I’ve removed it.
9d1c7e3bf05ce4ef8d405814706edd51f63a2347

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,341 @@
+        --topupconfig="$TopupConfig" \
+        --bfsigma="$BiasFieldSmoothingSigma" \
+        --brainextract="$BrainExtract" \
+        --t2wtype="$T2wType" \
```
I fixed it in 876369d764db288834797bc62d56601346c97d67 .

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+#T2wTemplate2mm="${HCPPIPEDIR_Templates}/MNI152_T2_2mm" 
+#TemplateMask="${HCPPIPEDIR_Templates}/MNI152_T1_${StrucRes}mm_brain_mask.nii.gz" 
+#Template2mmMask="${HCPPIPEDIR_Templates}/MNI152_T1_2mm_brain_mask_dil.nii.gz" 
+#UsePhaseZero="FALSE" #Indicates whether to add T2-weighted image as a phase zero volume, for dark-CSF T2w contrast acquisition types (e.g., FLAIR)
```
This depends on what type of T2w scan was used (and possibly on whether topup is struggling), not on what species it is, so I don't think it belongs in `SetUpSPECIES.sh`.

glasserm replied:
Looks like we all agree on that.

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+                SpinEchoPhaseEncodePositive2=${StudyFolder}/${Subject}/RawData/$(echo $StrucTopupPositive2 | awk '{print $1}') # first SEfield will be used
+            fi
+            SEEchoSpacing="$(echo $StrucSEDwellTime | awk '{print $1}')"    # dwell time for SE field, read from SEEchoSpacing in Seriesinfo.txt
+            SEUnwarpDir="$(echo $StrucSEUnwarpDir | awk '{print $1}')"  # read from PhaseEncodinglist in Seriesinfo.txt. x or y (minus or not does not matter)
```
Again, making a new variable with a more ambiguous name and `awk`ing for the first item are strange design choices to me.

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -318,6 +342,9 @@ log_Msg "Parsing Command Line Options"
 # Convert UseJacobian value to all lowercase (to allow the user the flexibility to use True, true, TRUE, False, False, false, etc.)
 UseJacobian="$(opts_StringToBool ${UseJacobian})"
 
+# Set UsePhaseZero to true or false
```
As long as we have a comment here, let's make it accurate to what `opts_StringToBool` does.
```suggestion
# Set UsePhaseZero to 1 or 0
```

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +503,491 @@ fi
 log_Msg "Completed!"
 
+#### Next stage: FreeSurfer/FreeSurferPipeline.sh
+
```
I don't know who this comment is intended for, I would remove it.

coalsont commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -27,24 +27,71 @@ opts_SetScriptDescription "Tool for creating a 6 DOF alignment of the AC, ACPC l
 
 #optional args
-opts_AddOptional '--ref' 'Reference' 'image' 'reference image' "${FSLDIR}/data/standard/MNI152_T1_1mm"
+opts_AddOptional '--ref' 'Reference' 'image' 'reference image' "${FSLDIR}/data/standard/MNI152_T1_1mm.nii.gz"
```
Adding the `.nii.gz` to the default is a good catch, as it has needed to work in `wb_command -convert-affine` since 2023 (https://github.com/Washington-University/HCPpipelines/commit/5719c91af6698445b0cc632830e37aef044a4c32).  We don't actually call this script without an explicit `--ref` parameter, so we didn't find this bug.

Defaulting to something in `$FSLDIR` is a bit odd (was added with the old-style parsing by a big "Oxford review" commit in 2013, https://github.com/Washington-University/HCPpipelines/commit/f5e3b5818c054aaedea5e3915ead30e5e75a6240), we could consider converting `--ref` to mandatory.  @glasserm ?

TakJim replied:
I made --ref mandatory.
59c5b8ba685e7d5e7eb8586e18af3eebc7d05ac6

coalsont commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -120,61 +161,205 @@ fi
+    Phase2ndDir=TRUE
+else
+    Phase2ndDir=FALSE
+fi
```
The `log_Msg` lines added here are redundant with `opts_ShowValues`, this is why none of the other parameters get logged manually.

```suggestion

if [[ ! -z $PhaseEncodeOne2 && $PhaseEncodeOne2 != NONE ]] ; then 
    if [[ `fslhd $PhaseEncodeOne2 | grep '^dim[123]'` != `fslhd $ScoutInputName | grep '^dim[123]'` ]]
    then
        log_Err_Abort "2nd pair of spin echo fieldmap images have different dimensions!"
    fi
    ${FSLDIR}/bin/imcp $PhaseEncodeOne2 ${WD}/PhaseOne2
    ${FSLDIR}/bin/imcp $PhaseEncodeTwo2 ${WD}/PhaseTwo2
    Phase2ndDir=TRUE
else
    Phase2ndDir=FALSE
fi
```

TakJim replied:
Thank you for your suggestion. I've fixed it.
374f6b1b2192d35b516a84898778587745f988d8

coalsont commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -55,13 +55,53 @@ opts_AddOptional '--topupconfig' 'TopupConfig' 'path' "topup config file"
+opts_AddOptional '--scannerpatientposition' 'ScannerPatientPosition' 'string' "HFS (default), HFP"
+
+opts_AddOptional '--truepatientposition' 'TruePatientPosition' 'string' "HFS (default), HFSx, FFSx"
+
```
If we say something is the default, let's actually have it as the default.

```suggestion
opts_AddOptional '--scannerpatientposition' 'ScannerPatientPosition' 'string' "HFS (default), HFP" 'HFS'

opts_AddOptional '--truepatientposition' 'TruePatientPosition' 'string' "HFS (default), HFSx, FFSx" 'HFS'

```

coalsont commented on global/scripts/CorrectVolumeOrientation.sh:
```diff
@@ -0,0 +1,236 @@
+opts_AddMandatory '--sposition'        'PatientPosition' 'HFS, HFP, FFS, FFP, NONE'    "patient position entered in the scanner, described in the tag (0018,5100) in the standard DICOM. If specified NONE, it does not change sform unless specified --init option."
+opts_AddMandatory '--out'              'out'             'volume'                 "output volume filename." 
+opts_AddOptional  '--omat'             'omat'            'TRUE or NONE (default)' "output transformation matrix for reorientation. Outputs are <out>_reorient.world.mat (in world format) and <out>_reorient.mat (FSL flirt format)." "NONE"
+opts_AddOptional  '--reorient2std'     'reorient2std'    'TRUE (default) or NONE' "reorient to RPI ('radiological') convention." 'TRUE'
```
Use `opts_StringToBool` for this also, to catch typos.

TakJim replied:
I fixed it.
2583334a80de6b669aadb0dc5ea3b5b2f119dc33

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+        --topupconfig=${TopupConfig} \
+        --bfsigma=${BiasFieldSmoothingSigma} \
+        --brainextract=${BrainExtract} \
+        --t2wtype=${T2wType} \
```
fixed it in c235f246fac2a5e654f42ae01f31261510d28016.

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+        if [[ -n $StrucTopupPositive && -n $StrucTopupNegative && -n $StrucSEDwellTime && "$StrucTopupPositive" != NONE ]] ; then
+            SpinEchoPhaseEncodeNegative=${StudyFolder}/${Subject}/RawData/$(echo $StrucTopupNegative | awk '{print $1}') # first SEfield will be used
+            SpinEchoPhaseEncodePositive=${StudyFolder}/${Subject}/RawData/$(echo $StrucTopupPositive | awk '{print $1}') # first SEfield will be used
+            if [[ -n $StrucTopupPositive2  && -n $StrucTopupNegative2 && "$StrucTopupPositive2" != NONE ]] ; then
```
I determined that set -eu is unnecessary in SetUpSpecies.sh, so I removed it.
https://github.com/Washington-University/HCPpipelines/commit/f57aaf42cd69f37358fc09954c71e9739c4b4fec

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+    if [[ $T1wSampleSpacing != "" && $T1wSampleSpacing != "NONE" && $T1wSampleSpacing != "None" && $T2wSampleSpacing != "" && $T2wSampleSpacing != "NONE" && $T2wSampleSpacing != "None" && $StrucSEUnwarpDir != "None" && $StrucSEUnwarpDir != "" ]] ; then
+        T1wSampleSpacing=$T1wSampleSpacing  # read from Seriesinfo.txt
+        T2wSampleSpacing=$T2wSampleSpacing  # read from Seriesinfo.txt
+        UnwarpDir="$StrucUnwarpDir"         # Read direction of structural scan. read from Seriesinfo.txt
```
fixed it in efdeb9995b526c1f2ec23fc86877331c607f2836.

TakJim commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -395,11 +411,53 @@ for TXw in $Modalities ; do
+                ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_warped${TXw}2${TXwImageBasename}_init -ref ${TXwImage} -out ${WD}/Magnitude_warped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}_tmp.mat -finesearch 2
+                ${FSLDIR}/bin/convert_xfm -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -concat ${WD}/Fieldmap2${TXwImageBasename}_tmp.mat  ${WD}/Fieldmap2${TXwImageBasename}_init.mat
+                rm -f ${WD}/Fieldmap2${TXwImageBasename}_tmp.mat                        
+            fi
```
It may be better to stick with the typo name for compatibility. I have also updated the conditional logic.
ba7b592f398d45a84669dcb90a7026b2ba9ff4a1

TakJim commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -55,13 +55,53 @@ opts_AddOptional '--topupconfig' 'TopupConfig' 'path' "topup config file"
+    log_Err_Abort "Invalid species: '$SPECIES'. Must be one of: Human, Macaque, Rhesus, Chimp, NightMonkey, Marmoset."
+    ;;
+esac
+
```
That makes sense. I fixed it.
bbbaca00df2a9acc20f793f27291ecb7d43757e7

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+# species specific config
+# If StrucRes is not provided, SetUpSPECIES.sh will use species-specific default
+#SPECIES="Human"
+#StrucRes=0.8 # 0.8 or 0.7 for Human, 0.5 or 0.3 for Macaque, 0.2 for Marmoset.
```
Please also move the `# 0.8 or 0.7 for Human, 0.5 or 0.3 for Macaque, 0.2 for Marmoset.` comment up to the `StrucRes="0.8"` line.

takuya-hayashi commented on global/scripts/CorrectVolumeOrientation.sh:
```diff
@@ -0,0 +1,236 @@
+  echo " 0  0 -1  0" >> ${tmp}_reorient_I.world.mat
+  echo " 0  1  0  0" >> ${tmp}_reorient_I.world.mat
+  echo " 0  0  0  1" >> ${tmp}_reorient_I.world.mat
+else
```
Thanks. Now I understand it and cofirmed it by the data.

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+        bettop2center="30"           # distance in mm from the top of FOV to the center of brain in robustroi
+
+        BrainTemplate="Mac30BS"
+        T1wTemplate="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/MNINonLinear/${BrainTemplate}_T1w_restore.nii.gz" 
```
If Mac30BS will never have another resolution, then maybe we should check that `$StrucRes` equals whatever the resolution is in this template.  Otherwise, we may want to put `$StrucRes` into the naming of these templates for consistency.

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+        bettop2center="30"    
+
+        BrainTemplate="Mac6Snow"
+        T1wTemplate="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/MNINonLinear/${BrainTemplate}_T1w_restore.nii.gz"
```
Also doesn't have `$StrucRes`.

coalsont commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+    TopupConfig="${HCPPIPEDIR_Config}/b02b0_marmoset_fMRI.cnf" #Config for topup or "NONE" if not used
+    BiasFieldSmoothingSigma="2.5"
+    BrainTemplate="NightMonkey9"
+    T1wTemplate="${HCPPIPEDIR_Templates}/NHP_NNP/${BrainTemplate}/MNINonLinear/${BrainTemplate}_T1w_restore_0.25mm.nii.gz"
```
This hardcodes the equivalent of `$StrucRes`.

TakJim commented on global/scripts/TopupPreprocessingAll.sh:
```diff
@@ -120,61 +161,205 @@ fi
+        ${FSLDIR}/bin/imcp ${WD}/PhaseTwo2 ${WD}/PhaseTwo2_gdc
+    fi
+
+    if [[ ("$TruePatientPosition" = "HFSx" || "$TruePatientPosition" = "FFSx" || "$TruePatientPosition" = "HFS" || "$TruePatientPosition" = "FFS" ) && ( "$TruePatientPosition" != "$ScannerPatientPosition") ]] ; then
```
Thanks for the suggestion. I've fixed it.
4bd16863b6b3df8b8b3598ca0bb5c3ab272194e8

TakJim commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -395,11 +411,55 @@ for TXw in $Modalities ; do
             ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_warpped${TXw} -ref ${TXwImage} -out ${WD}/Magnitude_warpped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
+            if [[ "$SPECIES" != "Human" ]] ; then
+                if [[ ! $SPECIES == *Marmoset* ]] ; then
+                    ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_warpped${TXw} -ref ${TXwImage}.nii.gz -out ${WD}/Magnitude_warpped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
```
I fixed the branching between Human and NonHuman.
86cb88fe1333228ec3d0474eb36b0719435dbf88

TakJim replied:
I simplified the conditional branch.
b52956c847b77c2e65ce913e9aaf5220e2dc7fd8

coalsont commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -394,12 +410,57 @@ for TXw in $Modalities ; do
-            ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_brain_warpped${TXw} -ref ${TXwImageBrain} -out ${WD}/Magnitude_brain_warpped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
+            if [[ ! $SPECIES == *Marmoset* ]] ; then
+                ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/Magnitude_brain -r ${WD}/Magnitude_brain -w ${WD}/FieldMap_Warp${TXw}.nii.gz -o ${WD}/Magnitude_brain_warped${TXw}
+                ${FSLDIR}/bin/flirt -interp spline -dof 6 -in ${WD}/Magnitude_brain_warped${TXw} -ref ${TXwImageBrain} -out ${WD}/Magnitude_brain_warped${TXw}2${TXwImageBasename} -omat ${WD}/Fieldmap2${TXwImageBasename}.mat -searchrx -30 30 -searchry -30 30 -searchrz -30 30
```
This also looks like "warpped" typo fixing that we decided against.

TakJim replied:
I reverted it.
6bf9cff631893665c525a288b4c3cd38b853dc54

coalsont commented on global/scripts/CorrectVolumeOrientation.sh:
```diff
@@ -0,0 +1,236 @@
+readsform ${in} ${tmp}_sform.mat
+convert_xfm -omat ${tmp}_reorient.world.mat -concat ${tmp}_reorient_II.world.mat ${tmp}_reorient_I.world.mat
+convert_xfm -omat ${tmp}_newsform.mat -concat ${tmp}_reorient.world.mat ${tmp}_sform.mat
+${CARET7DIR}/wb_command -volume-set-space ${in} ${tmp}.nii.gz -sform $(cat ${tmp}_newsform.mat | head -3)
```
As I said, the matrices may need to be changed if they used to work with the old (non-inverse, reverse order) logic.  I'm not sure how much change is required to get them working again.

With a closer look, `--tposition=HFSx` is already an identity matrix (head first sphinx, I assume), which is a fairly intuitive convention for bore coordinates to me, so you can use that with a conventional-orientation image as input to debug all the `--sposition` settings (and then copy the matrices to the `--tposition` section), as those settings should transform the input from patient-oriented coordinates (which the input file already is in) to what the bore coordinates would have been if the monkey and scanner settings were both equal to the specified orientation (compare it to the original image to check whether it matches how you would "see" the monkey relative to sphinx by looking at the scanner from the control booth).

Based on that, the current FFSx matrix suggests the monkey's head is upside down but looking out the same side of the bore as HFSx, and I don't know enough to say whether that is wrong (literal "feet first" would imply that FFSx is now looking towards its feet, though, which raises some mechanical questions).

TakJim commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,342 @@
+        T2wSampleSpacing=$T2wSampleSpacing  # read from Seriesinfo.txt
+        UnwarpDir="$StrucUnwarpDir"         # Read direction of structural scan. read from Seriesinfo.txt
+        if [[ -n $StrucTopupPositive && -n $StrucTopupNegative && -n $StrucSEDwellTime && "$StrucTopupPositive" != NONE ]] ; then
+            SpinEchoPhaseEncodeNegative=${StudyFolder}/${Subject}/RawData/$(echo $StrucTopupNegative | awk '{print $1}') # first SEfield will be used
```
I removed unnecessary awk statement. b72517d9c72226d1fa4c0daff8b8499fb40e4140

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,337 @@
+    if [[ $Gradient = "" || $Gradient = "NONE" || $Gradient = "None" ]] ; then
+        GradientDistortionCoeffs="NONE"               #Location of Coeffs file or "NONE" to skip
+    else
+        GradientDistortionCoeffs=${HCPPIPEDIR_Config}/coeff_${Gradient}.grad
```
Let's just make this a full absolute path to the gradient coefficents without an extra variable.

takuya-hayashi reviewed (commented):
Updated CorrectVolumeOrientation.sh
[59e1f9f](https://github.com/Washington-University/HCPpipelines/pull/350/commits/59e1f9fe5c4cad867d5a8f9268cfed4aeb13f2ee)

TakJim commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+    betfraction="0.3"            # fractional intensity threshold for bet
+    bettop2center="86"           # Distance between top of FOV and center of brain
+
+    FNIRTConfig="${HCPPIPEDIR_Config}/T1_2_NHP_NNP_Human_2mm.cnf" #FNIRT 2mm T1w Config
```
I set FNIRTConfig=${FSLDIR}/etc/flirtsch/T1_2_MNI152_2mm.cnf.
677fa53be5da08c63f94ee2ac2306e604d9b5a09

TakJim commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -83,38 +130,91 @@ echo "date: `date`" >> $WD/log.txt
+  centerx=$(echo "$dim1*0.5" | bc | awk '{printf "%d", $1}')
+  centery=$(echo "$dim2*0.48" | bc| awk '{printf "%d", $1}')
+  centerz=$(echo "$dim3 - $BetTop2Center/$pixdim3" | bc | awk '{printf "%d", $1}') 
+  if [ "$BiasfieldCor" = TRUE ] ; then
```
I’ve updated the script so that BiasFieldCor is set and I also renamed the variable name to BetBiasFieldCor for clarity.
 d181ccf7a488fbfa52924cb64a1a9c4c49dcd033

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -275,6 +280,21 @@ opts_AddOptional '--custombrain' 'CustomBrain' 'NONE_or_MASK_or_CUSTOM' "If PreF
+opts_AddOptional '--brainextract' 'BrainExtract' 'string' "Brain extraction method (default: INVIVO)" "INVIVO"
+opts_AddOptional '--use-t2w-phase-zero' 'UsePhaseZero' 'TRUE/FALSE' "Indicates whether to add T2-weighted image as a phase zero volume, for bright-CSF T2w contrast acquisition types (e.g., not FLAIR)" "FALSE"
+opts_AddOptional '--bias-field-sigma-no-T2w' 'BiasFieldSmoothingSigmaNoT2w' 'value' "Bias Field Smoothing Sigma for Bias Field Correction using T1w image only (only for NHP, default: 20)" "20"
+opts_AddOptional '--betbiasfieldcor' 'BetBiasFieldCor' 'TRUE/FALSE' "Indicates whether to correct bias field for BET (default: FALSE)" "FALSE"
```
Please use `opts_StringToBool` to sanity check this parameter for typos in both this pipeline and again in `ACPCAlignment.sh`.  It accepts yes/no, true/false, and 0/1 as input and is case insensitive, so you can just pass the 0/1 variable directly to `ACPCAlignment.sh`.

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+
+#HACK: work around the log tool name hack in SetUpSPECIES.sh
+#since debug.shlib will be active by default, set the log toolname back to the Batch script
+log_SetToolName "$(basename -- "$0")"
```
This needs to be after `SetUpSPECIES.sh`, not before.

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+# Edit these variables before running
+StudyFolder="${HOME}/projects/Pipelines_ExampleData"
+Subjlist="100307 100610"
+SPECIES="Human"
```
Subjlist="nhp_session1 nhp_session2"

coalsont commented on ICAFIX/README.md:
````diff
@@ -43,6 +43,15 @@ your compute environment. In particular, the `${FSL_FIXDIR}/settings.sh` file
+conda env create -f "$HCPPIPEDIR"/environment.yml [-n AutoRecleanEnv or -p /somewhere/you/want/the/conda/environment/installed ]
+```
+If using the native python version, please specify `--python-interpreter` with its absolute path, such as `--python-interpreter=/usr/bin/python3`.
+
````
I think something went wrong with the rebase, there shouldn't be changes to these folders in the PR.

AutioJ commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -83,38 +134,91 @@ echo "date: `date`" >> $WD/log.txt
+    BC="-B"
+  fi
+  verbose_echo " --> Run initial BET with options: -m -r $BetRadius -c $centerx $centery $centerz -f $BetFraction -z $betspecieslabel $BC"
+  ${FSLDIR}/bin/bet4animal "$WD"/robustroi.nii.gz "$WD"/robustroi_brain -m -r $BetRadius -c $centerx $centery $centerz -f $BetFraction -z $betspecieslabel $BC
```
If BetBiasFieldCor="FALSE" then BC does not exist. Added BC="" to move forward. Could be added using "else" in L183.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+    betradius="75"               # brain radius for bet
+    betbiasfieldcor="TRUE"
+    betfraction="0.3"            # fractional intensity threshold for bet
+    bettop2center="86"           # Distance between top of FOV and center of brain
```
Please define betbiasfieldcor for all species (total of 8 changes needed).

TakJim replied:
fixed it in be00d862d946da685ec46594e0b45a53947045e2.

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,335 @@
+    ScannerPatientPosition=${StrucScannerPatientPosition:-HFS}
+
+# ${FSLDIR}/bin/fsl_sub ${QUEUE} ${LOG} \
+    ${HCPPIPEDIR}/PreFreeSurfer/PreFreeSurferPipeline.sh \
```
Recommendation made to fix.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+    SmoothingFWHM="1.6" #Recommended to be roughly the voxel size
+    GrayordinatesResolution="1.6" #should be either 1 (7T) or 2 (3T) for human. 
+
+elif [[ "$SPECIES" == *Macaque* ]] ; then
```
Recommendation made to fix.

TakJim commented on PreFreeSurfer/scripts/ACPCAlignment.sh:
```diff
@@ -83,38 +134,91 @@ echo "date: `date`" >> $WD/log.txt
+  centerz=$(echo "$dim3 - $BetTop2Center/$pixdim3" | bc | awk '{printf "%d", $1}') 
+  if (($BetBiasFieldCor)) ; then
+    BC="-B"
+  fi
```
fixed it in ededb6ca186729b9d48e7513f8d2d7574a71f3ab.

glasserm commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,341 @@
+    SmoothingFWHM="1.6" #Recommended to be roughly the voxel size
+    GrayordinatesResolution="1.6" #should be either 1 (7T) or 2 (3T) for human. 
+
+elif [[ "$SPECIES" == *Macaque* ]] ; then
```
We don't use that any more though.  I wonder if we should require rhesus or cyno.

AutioJ commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,336 @@
+
+    if [ -z "$GradientDistortionCoeffs" ] ; then
+        GradientDistortionCoeffs="NONE"               #Location of Coeffs file or "NONE" to skip
+
```
fi missing

glasserm replied:
```suggestion
fi
```

glasserm commented on Examples/Scripts/PreFreeSurferPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,336 @@
+# Edit these variables before running
+StudyFolder="${HOME}/projects/Pipelines_ExampleData"
+Subjlist="nhp_session1 nhp_session2"
+SPECIES="Macaque"
```
Change to something more specific, like RhesusMacaque.  Also comment here to select the species most similar to RhesusMacaque, CynoMacaque, SnowMacaque.

TakJim replied:
I set default $SPECIES RhesusMacaque.
8f237153eb11a7f7f580d3a82ceaefb1b3f5b20f

coalsont commented on global/scripts/CorrectVolumeOrientation.sh:
```diff
@@ -0,0 +1,260 @@
+
+if [ $reorient2std == NONE ] ; then
+  reorient2std=FALSE
+fi
```
Please don't allow "NONE" to be accepted as "false".  Remove this, and change the help info to not mention "NONE".

"NONE" is a holdover from positional arguments, and is generally a stand-in for `""`.  We should avoid supporting it in new scripts (unless there is a high likelihood of copy/paste of settings from something that currently supports it).

TakJim replied:
I removed this conditional check and change the value of reorient2std from "NONE" to "FALSE"
684e8496e281ce1752c3e54a2c2d020d30086ea5

TakJim commented on Examples/Scripts/SetUpSPECIES.sh:
```diff
@@ -0,0 +1,355 @@
+    bettop2center="12"           # distance in mm from the top of FOV to the center of brain in robustroi
+
+    FNIRTConfig="${HCPPIPEDIR_Config}/T1_2_NHP_NNP_Marmoset_0.4mm.cnf" #FNIRT 2mm T1w Config
+    TopupConfig="${HCPPIPEDIR_Config}/b02b0_marmoset_fMRI.cnf" #Config for topup or "NONE" if not used
```
I added for config files other species except for chimp.
a6ccdd20006b8814bb9934dfe816c177c0886259
@takuya-hayashi 
I couldn’t find the chimp .cnf in the BCIL environment. Do you know of any other location where it might be available?

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +507,491 @@ fi
+                                ${RUN} ${FSLDIR}/bin/fslreorient2std ${Image}_brain ${TXwFolder}/${TXwImage}${i}_gdc_brain
+                            fi
+                        fi
+                    fi
```
More "silently do nothing" tests for unlisted orientations.
```suggestion
                        if [[ "$TruePatientPosition" != "$ScannerPatientPosition" ]] ; then
                            log_Msg "Reorient $TruePatientPosition data with a scanner orientation of $ScannerPatientPosition"
                            ${RUN} ${HCPPIPEDIR_Global}/CorrectVolumeOrientation.sh --in=${Image} --out=${TXwFolder}/${TXwImage}${i}_gdc --tposition="$TruePatientPosition" --sposition="$ScannerPatientPosition"
                        else
                            ${RUN} ${FSLDIR}/bin/fslreorient2std $Image ${TXwFolder}/${TXwImage}${i}_gdc
                        fi 

                        if [ $(${FSLDIR}/bin/imtest $(remove_ext $Image)_brain) = 1 ] ; then # TH 2016 for ACPC initialization
                            log_Msg "Found $(remove_ext $Image)_brain"
                            if [[ $(${FSLDIR}/bin/imtest ${TXwFolder}/${TXwImage}${i}_gdc_brain) = 1 ]] ; then
                                ${RUN} ${FSLDIR}/bin/imrm ${TXwFolder}/${TXwImage}${i}_gdc_brain
                            fi
                            if [[ "$TruePatientPosition" != "$ScannerPatientPosition" ]] ; then
                                ${RUN} ${HCPPIPEDIR_Global}/CorrectVolumeOrientation.sh --in=${Image}_brain --out=${TXwFolder}/${TXwImage}${i}_gdc_brain --tposition="$TruePatientPosition" --sposition="$ScannerPatientPosition"
                            else
                                ${RUN} ${FSLDIR}/bin/fslreorient2std ${Image}_brain ${TXwFolder}/${TXwImage}${i}_gdc_brain
                            fi
                        fi
                    fi
```

TakJim replied:
Fixed it in 1e603c61.

TakJim commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -415,354 +507,491 @@ fi
+                    --owarp=${TXwFolder}/xfms/${TXwImage}${i}_gdc_warp
+                    
+                    if [ "$SPECIES" != "Human" ] ; then
+                        if [[ ("$TruePatientPosition" = "HFSx" || "$TruePatientPosition" = "FFSx" || "$TruePatientPosition" = "HFS" || "$TruePatientPosition" = "FFS" ) && ( "$TruePatientPosition" != "$ScannerPatientPosition") ]] ; then
```
Fixed it in 346050f2.

coalsont commented on global/scripts/CorrectVolumeOrientation.sh:
```diff
@@ -0,0 +1,257 @@
+  --sposition=HFP --ref=T1w_d1_reorient.nii.gz --out=T1w_d2_reorient.nii.gz
+$ CorrectVolumeOrientation --in=fMRI_d2_HFSx.nii.gz --tposition=HFSx 
+  --sposition=HFP --init=T1w_d2_reorient2ref.world.mat 
+  --out=fMRI_d2_reorient2ref.nii.gz
```
`.sh` in examples:
```suggestion
$ CorrectVolumeOrientation.sh --in=MRI.nii.gz --tposition=HFSx --sposition=HFP 
  --out=MRI2reorient.nii.gz

Ex) #2. structural MRI scanned in true position of HFS on day1 (T1w_d1_HFS) and 
fMRI scanned in true position in HFSx on day2 (fMRI_d2_HFSx) and additional fast 
T1w scanned for registration purpose on day2 (T1w_d2_HFSx), where scanner's 
patient orientation is set to HFS on day1 and HFP on day2. In this case, the 
fast T1w scanned on day2 is corrected for the head orientation and registered to 
the T1w on day1, then the fMRI data on day2 is corrected for head orientation 
and initialized for head position using fast T1w scanned on day2. In this case, 
the option --ref is useful to 'fake' the output volume as if it were scanned on 
the same position as in day1.

$ CorrectVolumeOrientation.sh --in=T1w_d2_HFSx.nii.gz --tposition=HFSx 
  --sposition=HFP --ref=T1w_d1_HFS.nii.gz --out=T1w_d2_reorient.nii.gz
$ CorrectVolumeOrientation.sh --in=fMRI_d2_HFSx.nii.gz --tposition=HFSx 
  --sposition=HFP --init=T1w_d2_reorient2ref.world.mat 
  --out=fMRI_d2_reorient2ref.nii.gz

Please confirm that the registration is successful by viewing T1w_d1 and 
T1w_d2_reorient2ref. If this is not the case, please goto Ex) #4.

Ex) #3. structural MRI on day1, function MRI and fast structural scan on day 2 
all scanned in true position of HFSx, where scanner's patient orientation is set 
to HFP. In this case, the structural MRI scan in day1 must be corrected for the 
head orientation first, then the structural and functional MRI in day2 are 
treated as in #1.

$ CorrectVolumeOrientation.sh --in=T1w_d1_HFSx.nii.gz --tposition=HFSx 
  --sposition=HFP --out=T1w_d1_reorient.nii.gz
$ CorrectVolumeOrientation.sh --in=T1w_d2_HFSx.nii.gz --tposition=HFSx 
  --sposition=HFP --ref=T1w_d1_reorient.nii.gz --out=T1w_d2_reorient.nii.gz
$ CorrectVolumeOrientation.sh --in=fMRI_d2_HFSx.nii.gz --tposition=HFSx 
  --sposition=HFP --init=T1w_d2_reorient2ref.world.mat 
  --out=fMRI_d2_reorient2ref.nii.gz
```

TakJim replied:
Fixed it in [fce096c](https://github.com/Washington-University/HCPpipelines/pull/350/commits/fce096cc56790381f9a4a90e6038508d1f7e2af3).
